src/charm.py

```diff
@@ -0,0 +1,477 @@
+            event.add_status(ops.BlockedStatus("example.com is not a valid main domain"))
+
+        # Do we have the required storage?
+        if self.model.storages:
+            if "media-cache" not in self.model.storages:
+                event.add_status(ops.WaitingStatus("Waiting for media-cache"))
+            if self.config["media-uri"] == "local://" and "local-media" not in self.model.storages:
+                event.add_status(ops.WaitingStatus("Waiting for local-media"))
+        else:
+            event.add_status(ops.WaitingStatus("Waiting for storage"))
+
+        # Is the peer relation configured?
+        if self.peers:
+            # Do we have the required secrets?
+            try:
+                secret_id = self.peers.data[self.app]["secret-id"]
+            except KeyError:
+                event.add_status(ops.BlockedStatus("Waiting for peer secrets"))
+            else:
+                try:
+                    self.model.get_secret(id=secret_id).get_content()
+                except ops.SecretNotFoundError:
+                    event.add_status(ops.BlockedStatus("Waiting for secret key"))
+        else:
+            event.add_status(ops.WaitingStatus("Waiting for peer relation"))
+
+        # Check if we have everything we need to connect to the database.
+        if not self.dsn:
+            event.add_status(ops.BlockedStatus("Waiting for database relation"))
+
+        # Is ingress configured?
+        if not self.ingress.is_ready():
+            event.add_status(ops.WaitingStatus("Waiting for ingress."))
+
+    @staticmethod
+    def _generate_secret_key(length=128):
+        """Return a random sequence of characters suitable for a secret key."""
+        choice = random.SystemRandom().choice
+        return "".join(choice(string.ascii_uppercase + string.digits) for _ in range(length))
+
+    def _on_install(self, event: ops.InstallEvent):
+        if not self.unit.is_leader():
+            logger.debug("Non-leader doesn't need to handle install.")
+            return
+        if not self.peers:
+            logger.debug("Waiting for peer relation before creating secret key.")
+            return
+        self._add_secret_key()
+
+    def _on_peer_relation_created(self, event: ops.RelationCreatedEvent):
+        if not self.unit.is_leader():
+            logger.debug("Non-leader doesn't need to handle peer-relation-created.")
+            return
+        self._add_secret_key()
+
+    def _add_secret_key(self):
+        """Generate a secret key for Takahē and store it as a Juju secret."""
+        if not self.peers:
+            logger.warning("Cannot add secret key before peer relation exists.")
+            return
+        # This is used for internal cryptography.
+        secret_key = self._generate_secret_key()
+        secret = self.unit.add_secret({"takahe-secret-key": secret_key})
+        if secret.id is None:
+            # TODO: Can this be tightened up in the ops type hinting?
```

Comment:
I think it will be hard, because of [this](https://ops.readthedocs.io/en/latest/#ops.Secret.id): "This will be None if the secret was obtained using [Model.get_secret()](https://ops.readthedocs.io/en/latest/#ops.Model.get_secret) with a label but no ID."

Reply:
Ah, indeed, because `secret-get` only returns the contents, not any secret metadata.